### PR TITLE
Add route specific onRequest hook

### PR DESCRIPTION
**Pull request by jchip**
Hello, thanks for creating redbird. 

I'd like to use redbird for our project.  There are two things that prevent us from doing so:

1. We need an `onRequest` hook for each route so we can modify req.headers etc.  Similar to haproxy's reqadd/reqdel commands. https://cbonte.github.io/haproxy-dconv/1.7/configuration.html#4-reqadd
2. There's a bug that redbird is appending `/` to routes like `/test` and would get 404 when trying to forward it.

I also update to use prettier to format the code.

Thank you for your time and your consideration.

### CODE REVIEW

**Comment by jchip** [comment]
@manast  Hi, I'd like to get this merged.  Please let me know what other changes I can make to get this accepted.  Thanks.

File: lib/proxy.js [hunk 1]
```diff
@@ -590,65 +620,79 @@ ReverseProxy.buildRoute = function (route) {
...
-    //
-    if (target.useTargetHostHeader === true) {
-      req.host = target.host;
-    }
+      this.log &&
+        this.log.info('Proxying %s to %s', src + url, path.posix.join(target.host, req.url));
 
-    this.log && this.log.info('Proxying %s to %s', src + url, path.join(target.host, req.url));
+      if (route.opts.onRequest) {
+        const x = route.opts.onRequest(req, res, target);
```
Comment (manast) [inline lib/proxy.js]: x is not a very descriptive variable name, can you use a better name?

File: lib/proxy.js [hunk 2]
```diff
@@ -590,65 +620,79 @@ ReverseProxy.buildRoute = function (route) {
...
 
-    //
-    // Host headers are passed through from the source by default
-    // Often we want to use the host header of the target instead
-    //
-    if (target.useTargetHostHeader === true) {
-      req.host = target.host;
-    }
+      this.log &&
+        this.log.info('Proxying %s to %s', src + url, path.posix.join(target.host, req.url));
```
Comment (manast) [inline lib/proxy.js]: this log becomes incorrect since it shows the target.host however maybe we are goint to proxy to x

File: lib/proxy.js [hunk 3]
```diff
@@ -698,9 +740,9 @@ function redirectToHttps(req, res, target, ssl, log) {
   req.url = req._url || req.url; // Get the original url since we are going to redirect.
 
   var targetPort = ssl.redirectPort || ssl.port;
-  var hostname = req.headers.host.split(':')[0] + ( targetPort ? ':' + targetPort : '' );
-  var url = 'https://' + path.join(hostname, req.url);
-  log && log.info('Redirecting %s to %s', path.join(req.headers.host, req.url), url);
+  var hostname = req.headers.host.split(':')[0] + (targetPort ? ':' + targetPort : '');
+  var url = 'https://' + path.posix.join(hostname, req.url);
```
Comment (manast) [inline lib/proxy.js]: why is better to use posix in this case?

**Comment by manast** [comment]
good job. It is a bit more time consuming to review since you also prettified the code in the same PR, so I need to find the relevant changes from the pure formatting ones.

File: lib/proxy.js (hunk 3 above)
Comment (jchip) [reply lib/proxy.js]: On windows, `path.join` would use `\`.

**Comment by jchip** [comment]
yup, my bad on the formatting.  Let me reapplying these changes on latest w/o the formatting.

**Comment by jchip** [comment]
@manast thanks.  closing this one for https://github.com/OptimalBits/redbird/pull/247 that reapply the changes w/o prettifying.  

I'd be happy to help you out on reviewing and merging pending PRs etc.  I have a lot more fixes and enhancements that I'd like to get merged back to your version if possible.

**Pull request closed without merging**